### PR TITLE
NNCF helpers

### PR DESCRIPTION
**Pull request by KodiaqQ**
### Changes

- Added helpers module to NNCF.
- Added dataset helper.

### Reason for changes

- Extended data-free usage.

### Related tickets

- 152550

### Tests

- Added `tests/common/test_helpers.py`

### CODE REVIEW

**Comment by AlexKoff88** [comment]
I am ok with the code but two comments from my side:
- Why is it not a part of `nncf.data` namespace?
- Why do we need requirements file inside, especially with "-c constraints.txt"?

**Comment by KodiaqQ** [comment]
> * Why is it not a part of `nncf.data` namespace?

Since NNCF does not have `helpers` namespace, it was introduced. Should I introduce `nncf.data.helpers` module for a corresponded namespace?

> * Why do we need requirements file inside, especially with "-c constraints.txt"?

The provided code utilizes `torch` and `transformers` modules. They are not part of the main NNCF requirements, so I provided an external `requirements.txt` list for the helpers module.

**Comment by AlexKoff88** [comment]
> > * Why is it not a part of `nncf.data` namespace?
> 
> Since NNCF does not have `helpers` namespace, it was introduced. Should I introduce `nncf.data.helpers` module for a corresponded namespace?

I missed that part during the last review but I would like to name the function `generate_text_data(model: PreTrainedModel, tokenizer: PreTrainedTokenizer, max_length: int = 32,  num_samples: int = 128) -> List[str]`
It is up to you how to arrange it internally. I would like to see the import like this: `from nncf.data import generate_text_data`.

> 
> > * Why do we need requirements file inside, especially with "-c constraints.txt"?
> 
> The provided code utilizes `torch` and `transformers` modules. They are not part of the main NNCF requirements, so I provided an external `requirements.txt` list for the helpers module.

Ok, thanks.

**Comment by AlexanderDokuchaev** [comment]
@KodiaqQ @AlexKoff88 @alexsu52 
Is planning to add other helper function to nncf.helpers?
May be will be better to locate the function in `nncf.data` or `nncf/quantization/algorithms` ?

**Comment by KodiaqQ** [comment]
> function in `nncf.data` or `nncf/quantization/algorithms` ?

Please suggest the appropriate place for this function considering that this method requires `transformers` & `torch` Python packages and shouldn't be mixed with the `nncf.Dataset`.

File: nncf/helpers/dataset.py
Comment (alexsu52) [inline nncf/helpers/dataset.py:1]: I would suggest to move this file to `nncf/data/generators.py`

File: nncf/helpers/requirements.txt
Comment (alexsu52) [inline nncf/helpers/requirements.txt:1]: I would suggest to remove requirements.txt and implement lazy checking of installed packages. Many repositories use this approach for additional requirements, such as `transformers`, `optimum` and etc.

**Review by alexsu52 — CHANGES_REQUESTED** [review]

File: nncf/helpers/dataset.py
Comment (KodiaqQ) [reply nncf/helpers/dataset.py:1]: Why `generators`? How it would correlate with the requirement to have `from nncf.data import generate_text_data` import?

File: nncf/helpers/requirements.txt
Comment (KodiaqQ) [reply nncf/helpers/requirements.txt:1]: Could you please provide the proper example? As I understand, the current approach is not valid for you.

**Review by andreyanufr — APPROVED** [review]

File: nncf/helpers/requirements.txt
Comment (alexsu52) [reply nncf/helpers/requirements.txt:1]: Example: 
https://github.com/huggingface/optimum/blob/2fb5ea5ca7ca8ea887af2851cce80ab2545d3f4f/optimum/gptq/quantizer.py#L742

File: nncf/helpers/dataset.py
Comment (alexsu52) [reply nncf/helpers/dataset.py:1]: It does not influence to the requirement to have from `nncf.data import generate_text_data` import. Could you make a proposal taking account removing `helpers` folder?

Comment (KodiaqQ) [reply nncf/helpers/dataset.py:1]: No proposals anymore.

File: nncf/helpers/requirements.txt
Comment (KodiaqQ) [reply nncf/helpers/requirements.txt]: Done.

File: tests/common/test_dataset_generators.py
Comment (AlexanderDokuchaev) [inline tests/common/test_dataset_generators.py:22]: Please use fixture to set seed like https://github.com/openvinotoolkit/nncf/blob/develop/tests/torch/conftest.py#L256
to guaranty that it will not affect to another tests and seed will be set individually for each test

File: tests/common/requirements.txt
Comment (AlexanderDokuchaev) [inline tests/common/requirements.txt:2]: As it used torch and transformers, i suggest to add this test to tests/torch

File: nncf/data/generators.py
Comment (AlexanderDokuchaev) [inline nncf/data/generators.py:26]: It will be good to add to docstring  limitations for model and that function requires torch and transformers modules.

**Review by AlexanderDokuchaev — CHANGES_REQUESTED** [review]

File: nncf/data/generators.py
Comment (AlexanderDokuchaev) [inline nncf/data/generators.py:70]: This function will be used only for torch modules or for OV too?
May be will useful to add option to return numpy.array to avoid installation torch? (add argument pt_tensor)

**Review by AlexanderDokuchaev — CHANGES_REQUESTED** [review]

File: nncf/data/generators.py
Comment (KodiaqQ) [reply nncf/data/generators.py:70]: OV can pass torch tensor as the input too.
It is not possible to avoid torch installation due to `transformers.PreTrainedModel` being the input. 
See the source for this class - https://github.com/huggingface/transformers/blob/v4.44.2/src/transformers/modeling_utils.py#L1297, which contains `(nn.Module, ModuleUtilsMixin, GenerationMixin, PushToHubMixin, PeftAdapterMixin)` as the base classes.

File: nncf/data/generators.py
Comment (KodiaqQ) [reply nncf/data/generators.py:26]: Added.

File: tests/common/test_dataset_generators.py
Comment (KodiaqQ) [reply tests/common/test_dataset_generators.py]: Done.

File: tests/common/requirements.txt
Comment (KodiaqQ) [reply tests/common/requirements.txt]: Done. From now, we'll test the common code in `tests/torch`.

**Review by AlexanderDokuchaev — APPROVED** [review]

**Review by alexsu52 — COMMENTED** [review]
I would suggest to add an example of usage `generate_text_data` in `https://github.com/openvinotoolkit/nncf/tree/develop/examples/llm_compression`. It could be done in the follow-up PR, but you need to think about convenient use of the function in this PR.

File: nncf/data/generators.py
Comment (KodiaqQ) [reply nncf/data/generators.py:24]: @AlexKoff88 asked me to modify this in a few comments below:

> I missed that part during the last review but I would like to name the function generate_text_data(model: PreTrainedModel, tokenizer: PreTrainedTokenizer, max_length: int = 32,  num_samples: int = 128) -> List[str]

I changed as was asked. For now, you're requesting to roll back this change. Please, discuss this with @AlexKoff88 and provide your final decision on that.

Comment (MaximProshin) [reply nncf/data/generators.py:24]: @AlexKoff88 , can you please describe the use case with some example so that it's clear how the new API will be used? I see that there is some misunderstanding.

Comment (KodiaqQ) [reply nncf/data/generators.py:25]: @alexsu52, are there any updates on that? What is the final decision?

**Comment by KodiaqQ** [comment]
> I would suggest to add an example of usage `generate_text_data` in `https://github.com/openvinotoolkit/nncf/tree/develop/examples/llm_compression`. It could be done in the follow-up PR, but you need to think about convenient use of the function in this PR.

An example of the non-API usage method was added.

**Comment by alexsu52** [comment]
> > I would suggest to add an example of usage `generate_text_data` in `https://github.com/openvinotoolkit/nncf/tree/develop/examples/llm_compression`. It could be done in the follow-up PR, but you need to think about convenient use of the function in this PR.
> 
> An example of the non-API method usage was added.

Great example! Please add test for the example.

**Comment by KodiaqQ** [comment]
@AlexKoff88, re-review an example, please.

**Comment by KodiaqQ** [comment]
> Great example! Please add test for the example.

Done.

**Review by alexsu52 — APPROVED** [review]
LGTM

**Review by MaximProshin — APPROVED** [review]

**Review by AlexKoff88 — APPROVED** [review]

**Comment by KodiaqQ** [comment]
Since there are some issues with the `test_examples`, I propose to disable this example in tests until the issue is investigated and merge this functionality.
@alexsu52, @MaximProshin, what do you think?

**Comment by MaximProshin** [comment]
@KodiaqQ 

> Since there are some issues with the `test_examples`, I propose to disable this example in tests until the issue is investigated and merge this functionality. @alexsu52, @MaximProshin, what do you think?

I agree

**Comment by alexsu52** [comment]
> Since there are some issues with the `test_examples`, I propose to disable this example in tests until the issue is investigated and merge this functionality. @alexsu52, @MaximProshin, what do you think?

Good idea!

**Pull request merged**
